### PR TITLE
stats_over_http c++/ipmap updates

### PR DESCRIPTION
**Pull request by ezelkow1**
convert to c++ plugin
move ip allow setup to use ipmap
if a config is used link it to remap to initiate plugin reload on config change

### CODE REVIEW

**Comment by ezelkow1** [comment]
[approve ci clang-analyzer]

**Comment by ezelkow1** [comment]
[approve ci clang analyzer]

**Comment by bryancall** [comment]
@SolidWallOfCode is going to review this

**Comment by bryancall** [comment]
[approve ci]

File: doc/admin-guide/plugins/stats_over_http.en.rst [hunk 1]
```diff
@@ -41,9 +41,16 @@ default URL::
 
 where host and port is the hostname/IP and port number of the server.
 
+This plugin can also run as a :file:`remap.config` plugin::
+
+   map http://incoming/ http://outgoing/ @plugin=stats_over_http.so @pparam=(empty/undefined || desired stats URL path || location of config file)
```
Comment (SolidWallOfCode) [inline doc/admin-guide/plugins/stats_over_http.en.rst:46]: Is `@pparam` required in the empty case?

File: plugins/stats_over_http/stats_over_http.cc [hunk 2]
```diff
@@ -55,8 +57,8 @@
 #define SYSTEM_RECORD_TYPE (0x100)
 #define DEFAULT_RECORD_TYPES (SYSTEM_RECORD_TYPE | TS_RECORDTYPE_PROCESS | TS_RECORDTYPE_PLUGIN)
 
-#define DEFAULT_IP "0.0.0.0"
-#define DEFAULT_IP6 "::"
+#define DEFAULT_IP "0.0.0.0/0"
```
Comment (SolidWallOfCode) [inline plugins/stats_over_http/stats_over_http.cc:60]: Use C++ constructs. Something like
```
constexpr std::string_view DEFAULT_IP = "0.0.0.0/0";
std::string const DEFAULT_IP = "0.0.0.0/0";
```

File: plugins/stats_over_http/stats_over_http.cc [hunk 3]
```diff
@@ -89,12 +91,9 @@ static bool wrap_counters    = false;
 
 typedef struct {
```
Comment (SolidWallOfCode) [inline plugins/stats_over_http/stats_over_http.cc:92]: No `typedef`!
```
struct config_t { ... };
```

File: plugins/stats_over_http/stats_over_http.cc [hunk 4]
```diff
@@ -574,22 +574,25 @@ stats_origin(TSCont contp ATS_UNUSED, TSEvent event ATS_UNUSED, void *edata)
...
+        !memcmp(path, config->stats_path.c_str(), config->stats_path.length()))) {
+    TSDebug(PLUGIN_NAME, "not this plugins path, saw: %.*s, looking for: %s", path_len, path, config->stats_path.c_str());
     goto notforme;
   }
 
-  const struct sockaddr *addr = TSHttpTxnClientAddrGet(txnp);
-  if (!is_ip_allowed(config, addr)) {
+  TSSkipRemappingSet(txnp, 1); // not strictly necessary, but speed is everything these days
+
+  addr = TSHttpTxnClientAddrGet(txnp);
```
Comment (SolidWallOfCode) [inline plugins/stats_over_http/stats_over_http.cc:588]: For real C++ style, don't declare this above with an initializer, declare it here initialized by `TSHttpTxnClientAddrGet`.
```
auto addr = TSHttpTxnClienAddrGet(txnp);
```
Not sure why you changed that from the original. If you wanted to be really cutting edge,  you'd do
```
if ( auto addr = TSHttpTxnClientAddrGet(txnp) ; !is _ip_allowed(config, addr)) {
// ....
```

File: plugins/stats_over_http/stats_over_http.cc [hunk 5]
```diff
@@ -656,7 +659,17 @@ stats_origin(TSCont contp ATS_UNUSED, TSEvent event ATS_UNUSED, void *edata)
   if (accept_encoding_field) {
     TSHandleMLocRelease(reqp, TS_NULL_MLOC, accept_encoding_field);
   }
-  TSHttpTxnReenable(txnp, reenable);
+
+  return reenable;
+}
+
+static int
+global_stats_origin_wrapper(TSCont contp ATS_UNUSED, TSEvent event ATS_UNUSED, void *edata)
```
Comment (SolidWallOfCode) [inline plugins/stats_over_http/stats_over_http.cc:667]: Don't use `ATS_UNUSED`, just omit the name entirely.

File: plugins/stats_over_http/stats_over_http.cc [hunk 6]
```diff
@@ -705,250 +718,209 @@ TSPluginInit(int argc, const char *argv[])
   /* Path was not set during load, so the param was not a config file, we also
     have an argument so it must be the path, set it here.  Otherwise if no argument
     then use the default _stats path */
-  if ((config_holder->config != NULL) && (config_holder->config->stats_path == 0) && (argc > 0) &&
+  if ((config_holder->config != NULL) && (config_holder->config->stats_path.empty()) && (argc > 0) &&
```
Comment (SolidWallOfCode) [inline plugins/stats_over_http/stats_over_http.cc:721]: `NULL` -> `nullptr`

File: plugins/stats_over_http/stats_over_http.cc [hunk 7]
```diff
@@ -705,250 +718,209 @@ TSPluginInit(int argc, const char *argv[])
...
-    return true;
+  config_holder_t *config_holder;
+  config_holder = new_config_holder(argc > 2 ? argv[2] : NULL);
+  *ih           = static_cast<void *>(config_holder);
+  /* Path was not set during load, so the param was not a config file, we also
+  have an argument so it must be the path, set it here.  Otherwise if no argument
+  then use the default _stats path */
+  if ((config_holder->config != NULL) && (config_holder->config->stats_path.empty()) && (argc > 2) &&
+      (config_holder->config_path == NULL)) {
+    config_holder->config->stats_path = TSstrdup(argv[2] + ('/' == argv[2][0] ? 1 : 0));
```
Comment (SolidWallOfCode) [inline plugins/stats_over_http/stats_over_http.cc:776]: Why not make `stats_path` a `std::string` and avoid calling not only `TSstrdup` but `TSfree`?

File: plugins/stats_over_http/stats_over_http.cc [hunk 8]
```diff
@@ -705,250 +718,209 @@ TSPluginInit(int argc, const char *argv[])
...
-  char ip_port_text_buffer[INET6_ADDRSTRLEN];
-
-  if (!ipStr) {
-    config->ip6Count = 1;
-    ip = config->allowIps6 = TSmalloc(sizeof(struct in6_addr) + 1);
-    inet_pton(AF_INET6, DEFAULT_IP6, ip);
-    ip[sizeof(struct in6_addr)] = 0;
-    return;
+  IpAddr min, max;
+  std::string delimiter = ",";
```
Comment (SolidWallOfCode) [inline plugins/stats_over_http/stats_over_http.cc:840]: No, it's allocating that every time!
```
constexpr std::string_view delimiter = ",";
```

File: plugins/stats_over_http/stats_over_http.cc [hunk 9]
```diff
@@ -705,250 +718,209 @@ TSPluginInit(int argc, const char *argv[])
...
+  size_t pos_start = 0, pos_end, delimiter_len = delimiter.length();
+  std::string token;
```
Comment (SolidWallOfCode) [inline plugins/stats_over_http/stats_over_http.cc:842]: `ats_ip_range_parse` takes a `string_view` - why aren't these `string_view`?

File: plugins/stats_over_http/stats_over_http.cc [hunk 10]
```diff
@@ -705,250 +718,209 @@ TSPluginInit(int argc, const char *argv[])
...
-    if (!tok2) {
-      mask = 128;
-    } else {
-      mask = atoi(tok2);
-    }
-    ip[sizeof(struct in6_addr)] = mask;
-    TSDebug(PLUGIN_NAME, "%d) adding netmask: %s/%d", i + 1, inet_ntop(AF_INET6, ip, ip_port_text_buffer, INET6_ADDRSTRLEN),
-            ip[sizeof(struct in6_addr)]);
-    i++;
+  while ((pos_end = ipstring.find(delimiter, pos_start)) != std::string::npos) {
```
Comment (SolidWallOfCode) [inline plugins/stats_over_http/stats_over_http.cc:859]: `TextView`? 
```
while (ipstring) {
  token = ipstring.take_prefix_at(',');
```

File: plugins/stats_over_http/stats_over_http.cc [hunk 11]
```diff
@@ -705,250 +718,209 @@ TSPluginInit(int argc, const char *argv[])
...
-      parseIps(config, p);
-    } else if ((p = strstr(buffer, "allow_ip6="))) {
-      p += strlen("allow_ip6=");
-      parseIps6(config, p);
+      config->stats_path = cur_line.substr(p); // nstr(strtok_r(p, " \n", &p));
+    } else if ((p = cur_line.find("record_types=") != std::string::npos)) {
+      TSDebug(PLUGIN_NAME, "parsing record types");
+      p += strlen("record_types=") - 1;
+      config->recordTypes = strtol(cur_line.substr(p).c_str(), NULL, 16);
+    } else if ((p = cur_line.find("allow_ip=") != std::string::npos)) {
```
Comment (SolidWallOfCode) [inline plugins/stats_over_http/stats_over_http.cc:906]: `((p = cur_line.find("allow_ip=")) != std::string::npos)`

**Comment by ezelkow1** [comment]
Just some notes for myself. After some discussions we want to:

- Remove the default of `_stats` when being used as a remap plugin, so that the remap itself becomes the endpoint
- Still allow setting of the path via config file when running as a remap

This would let you do 
`map http://mystats.com http://whatever.com @plugin=stats_over_http.so`
and just be able to hit stats without the _stats endpoint on a remap. It would still be optionally available via the config file though for things like ATC which would require this.

We still want to leave the global usage as is, allowing a config or if the file doesnt exist use the param as the path, or default to _stats just because global usage needs its own path. On a remap though it will be config file based or nothing at all

**Comment by ezelkow1** [comment]
Strike that, reverse it :)

So we are just taking out the remap usage all together. Much of how this plugin works is legacy implementations, so if we want a remap plugin that can serve stats it should be done as just a whole new plugin written properly and so that it can be chained with other plugins.

For now removing the new remap functionality. I will leave the existing rewrite though as we still need the proper IP parsing/map but for remap it should be a new plugin designed around using the ServerIntercept

**Comment by ezelkow1** [comment]
[approve ci rocky]

**Comment by randall** [comment]
See https://github.com/apache/trafficserver/pull/8797 for stats_over_http reimplemented as a ServerIntercept.

**Comment by ezelkow1** [comment]
We have been running with this in production for a good while now

**Review by SolidWallOfCode — APPROVED** [review]

**Pull request merged**